tools/terraform/fleetdm_client/fleetdm_client.go

```diff
@@ -0,0 +1,276 @@
+package fleetdm_client
+
+// This file gives us a nice API to use to call FleetDM's API. It's focused
+// only on teams.
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var prefix = "/api/v1/fleet"
+var teamPrefix = prefix + "/teams"
+
+type Team struct {
+	Name         string      `json:"name"`
+	ID           int64       `json:"id"`
+	Description  string      `json:"description"`
+	AgentOptions interface{} `json:"agent_options"`
+	Scripts      interface{} `json:"scripts"`
+	Secrets      []struct {
+		Secret  string    `json:"secret"`
+		Created time.Time `json:"created_at"`
+		TeamID  int       `json:"team_id"`
+	} `json:"secrets"`
+}
+
+type TeamCreate struct {
+	Name        string `json:"name,omitempty"`
+	Description string `json:"description,omitempty"`
+}
+
+type TeamGetResponse struct {
+	Team struct {
+		Team
+	} `json:"team"`
+}
+
+type TeamQueryResponse struct {
+	Teams []Team `json:"teams"`
+}
+
+// FleetDMClient is a FleetDM HTTP client that overrides http.DefaultClient.
+type FleetDMClient struct {
+	*http.Client
+	URL    string
+	APIKey string
+}
+
+// NewFleetDMClient creates a new instance of FleetDMClient with the provided
+// URL and API key.
+func NewFleetDMClient(url, apiKey string) *FleetDMClient {
+	return &FleetDMClient{
+		Client: http.DefaultClient,
+		URL:    url,
+		APIKey: apiKey,
+	}
+}
+
+// Do will add necessary headers and call the http.Client.Do method.
+func (c *FleetDMClient) do(req *http.Request, query string) (*http.Response, error) {
+	// Add the API key to the request header
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", c.APIKey))
+	req.Header.Add("Accept", `application/json`)
+	// Set the request URL based on the client URL
+	req.URL, _ = url.Parse(c.URL + req.URL.Path)
+	if query != "" {
+		req.URL.RawQuery = query
+	}
+	// Send the request using the embedded http.Client
+	return c.Client.Do(req)
+}
+
+// TeamNameToId will return the ID of a team given the name.
+func (c *FleetDMClient) TeamNameToId(name string) (int64, error) {
+	req, err := http.NewRequest(http.MethodGet, teamPrefix, nil)
+	if err != nil {
+		return 0, fmt.Errorf("failed to create GET request for %s: %w", teamPrefix, err)
+	}
+	query := fmt.Sprintf("query=%s", name)
+	resp, err := c.do(req, query)
+	if err != nil {
+		return 0, fmt.Errorf("failed to GET %s %s: %w", teamPrefix, query, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return 0, fmt.Errorf("failed to get team: %s %s", query, resp.Status)
+	}
+
+	var teamqry TeamQueryResponse
+	err = json.NewDecoder(resp.Body).Decode(&teamqry)
```

Comment:
The `json.Decoder` is a JSON-lines reader, it accepts a list of JSON values separated by newlines. This means that it would not return an error if the response JSON was something like `{"foo":"bar"} "oops"` because it would only decode the first value. For more robust processing, it would be better to do `b := io.ReadAll(resp.Body)` and use `json.Unmarshal` with those bytes - this ensures that the response as a whole is a single valid JSON value.

(if like me your initial reaction to this is that you'd rather not read the full bytes in memory and stream it instead - it's good to note that the Decoder's `Decode` method will _also_ load the full JSON value in memory, it's just the sequence of JSON values that are streamed).

Reply:
Huh, didn't know that. Fixed.